spring-rabbit/src/main/java/org/springframework/amqp/rabbit/retry/MessageKeyGenerator.java

```diff
@@ -31,10 +32,13 @@ public interface MessageKeyGenerator {
 	 * about assuming uniqueness of any element of the message, especially considering the requirement that it be
 	 * repeatable. A message id is ideal, but may not be present (AMQP does not mandate it), and the message body is a
 	 * byte array whose contents might be repeatable, but its object value is not.
+	 * <p>While returning {@code null} is allowed, this represents a faulty scenario that
+	 * will prevent retry operations to work correctly.
```

Comment:
In Spring Retry Dave say:
```
		Object generatedKey = defaultKey;
		if (this.keyGenerator != null) {
			generatedKey = this.keyGenerator.getKey(invocation.getArguments());
		}
		if (generatedKey == null) {
			// If there's a generator and he still says the key is null, that means he
			// really doesn't want to retry.
			return null;
		}
```

May it is better to turn it off in favor of non-null contract?
Since we are not relying on Spring Retry any more and I don't see why we would allow nullable path where such a `MessageKeyGenerator` injection would be for nothing in the `StatefulRetryOperationsInterceptor` configuration.

Reply:
I am not sure what you mean. I want 100% this code to not allow null but Spring AMQP does not allow it. Try to put an assert non null and run the tests and you'll see.

Reply:
I don't have the history of the projet to know for sure. It'd be nice if that wasn't allowed but I doubt that's going to be easy. 

Feel free to rephrase as you see fit. I tried to document with my limited knowledge of the project.